### PR TITLE
Generic/ArrayIndent: add XML documentation

### PR DESCRIPTION
**Pull request by rodrigoprimo**
## Description

This PR adds the XML documentation for the Generic.Arrays.ArrayIndent sniff.

I left some inline comments below highlighting decisions I made while creating this documentation that I'm not sure about.

## Suggested changelog entry

Add XML documentation for the Generic.Arrays.ArrayIndent sniff.

## Related issues/external references

Part of https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/148

## Types of changes

- [ ] Bug fix _(non-breaking change which fixes an issue)_
- [ ] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [x] Documentation improvement

## PR checklist

- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [ ] I have added tests to cover my changes.
- [ ] I have verified that the code complies with the projects coding standards.
- [ ] [Required for new sniffs] I have added XML documentation for the sniff.

### CODE REVIEW

File: src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml [hunk 1]
```diff
@@ -0,0 +1,101 @@
+<documentation title="Array Indent">
+    <standard>
+    <![CDATA[
+    The opening square bracket/parenthesis of a multi-line array must be indented at least to the same level of the start of the statement.
```
Comment (rodrigoprimo) [inline src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml:4]: I opted to use `opening square bracket/parenthesis` here and in a few other places as the sniff supports both array syntaxes. That is not super correct as in the old syntax, the opening token of an array is not just a parenthesis, but actually `array(`.

File: src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+<em>    [</em>
+        1,
+        2,
+    ];
+}
+        ]]>
+        </code>
+        <code title="Invalid: Opening square bracket/parenthesis of a multi-line array not indented at least to the same level of the start of the statement.">
+        <![CDATA[
+if ($condition) {
```
Comment (rodrigoprimo) [inline src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml:21]: Initially, I wanted to use the snippet below as the invalid example but the generator code trims the leading spaces before `$a` (https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/8974f2662ac60cc2d6a545eda35e41c6f0fb62e8/src/Generators/Text.php#L131) and that is why I opted to add an `if` statement.

```php
    $a =
[
        1,
        2,
    ];
```

File: src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml (hunk 1 above)
Comment (rodrigoprimo) [inline src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml:4]: For this `<standard>` block, I'm specifically mentioning multi-line arrays. It is the only error that could apply to a single-line array, but single-line arrays are not handled by this sniff. The example that I have in mind is:

```php
    $a = 
[1, 2];
```

Comment (jrfnl) [reply src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml:4]: What about using the phrase "open brace" (and "close brace" for the end, of course) ? Would that work ? What do you think ?

Comment (jrfnl) [reply src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml:4]: Just tested, that snippet would not trigger the sniff anyway, so we can forget about it.

File: src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    </standard>
+    <code_comparison>
+        <code title="Valid: Opening square bracket/parenthesis token of a multi-line array indented at least to the same level of the start of the statement.">
+        <![CDATA[
+if ($condition) {
+    $a =
+<em>    [</em>
+        1,
+        2,
+    ];
```
Comment (jrfnl) [inline src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml:15]: What about adding a second code sample here showing that having the opening brace on the same line as the assignment is also perfectly valid under this rule ?

File: src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+}
+        ]]>
+        </code>
+    </code_comparison>
+    <standard>
+    <![CDATA[
+    The elements of an array must be indented four spaces beyond the indentation level of the opening square bracket/parenthesis.
```
Comment (jrfnl) [inline src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml:33]: Okay, not sure how to rephrase this, but there are some things here which don't ring true/could be confusing:
* The indentation is only checked for the first line of each element (subsequent lines are not checked for indentation).
* "Four spaces beyond"- not sure how much margin that give to interpret this as "at least four spaces", while the rule is "exactly four spaces".
* "the indentation level of the opening square bracket/parenthesis" - it's not about the indentation/position of the brackets, but about the start of statement which includes the array opener.

Let's rethink the phrasing of this rule.

Note: this also means the Valid/Invalid phrasing may need updating.

File: src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        <![CDATA[
+$a = [
+    1,
+    2,<em>]</em>;
+        ]]>
+        </code>
+    </code_comparison>
+    <standard>
+    <![CDATA[
+    The closing square bracket/parenthesis must be indented four spaces less than the array elements.
```
Comment (jrfnl) [inline src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml:80]: Would this be clearer if the indentation was phrased to be related to  "same level of indentation as the start of the statement containing the array opener" (or something along those lines) ?

**Review by jrfnl — CHANGES_REQUESTED** [review]
@rodrigoprimo Sorry for my tardiness with this review.

Generally speaking the documentation is good, there is just some phrasing which could be misinterpreted/interpreted in different ways, which may be confusing . Let's have another look at how to make it as clear as possible.

I think with some small tweaks this can probably be finished quickly. Again, my apologies for the delay in the review.

File: src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml (hunk 4 above)
Comment (rodrigoprimo) [reply src/Standards/Generic/Docs/Arrays/ArrayIndentStandard.xml]: I went with "The first line of each array element must be indented exactly four spaces more than the start of the statement". I also updated the valid/invalid sentences accordingly and changed the `<em>` tags to highlight only the spaces and not the array element. What do you think?

**Review by rodrigoprimo — COMMENTED** [review]
Thanks for your review, @jrfnl. I believe I addressed all the points that you raised. Could you please take a second look?

**Comment by jrfnl** [comment]
I've made some final tweaks to the docs after discussing these with @rodrigoprimo off-GH. Merging once the build has passed.

**Review by jrfnl — APPROVED** [review]

**Pull request merged**